neuraxle/metaopt/random.py

```diff
@@ -319,44 +323,54 @@ class RandomSearch(MetaStepMixin, BaseStep):
 
     def __init__(
             self,
+            wrapped,
             n_iter: int,
             higher_score_is_better: bool,
             validation_technique: BaseCrossValidation = KFoldCrossValidation(),
-            refit=True
+            refit=True,
+            print=False
     ):
+        MetaStepMixin.__init__(self, wrapped)
         BaseStep.__init__(self)
+        self.print = print
         self.n_iter = n_iter
         self.higher_score_is_better = higher_score_is_better
         self.validation_technique: BaseCrossValidation = validation_technique
         self.refit = refit
 
     def fit(self, data_inputs, expected_outputs=None) -> 'BaseStep':
-        # TODO: assert that set_step was called.
         started = False
+        best_hyperparams = None
+
         for _ in range(self.n_iter):
 
-            step = copy.copy(self.step)
+            step = copy.copy(self.wrapped)
 
             new_hyperparams = step.get_hyperparams_space().rvs()
             step.set_hyperparams(new_hyperparams)
 
             step: BaseCrossValidation = copy.copy(self.validation_technique).set_step(step)
 
-            # TODO: skip on error???
             step = step.fit(data_inputs, expected_outputs)
             score = step.scores_mean
 
             if not started or self.higher_score_is_better == (score > self.score):
                 started = True
                 self.score = score
-                self.best_validation_wrapper_of_model = step
+                self.best_validation_wrapper_of_model = copy.copy(step)
+                best_hyperparams = new_hyperparams
+
+        self.best_validation_wrapper_of_model.wrapped.set_hyperparams(best_hyperparams)
 
         if self.refit:
-            self.best_model = self.best_validation_wrapper_of_model.step.fit(
-                data_inputs, expected_outputs)
+            self.best_model = self.best_validation_wrapper_of_model.fit(
```

Comment:
Is this tested? I'd guess this would crash. The `.step` is replaced here by nothing like ` `, it should probably have been replaced by `.wrapped`.

Reply:
Ah, I see what you did here. You fit the wrapper instead of the wrapee. Here we want to fit the wrapee on everything, not the wrapper. Because we want to train on the validation set too when we use self.refit : we retrain on everything, so this must call the wrapped obj and not the wrapper that contains it. It should even copy the wrapped before fitting it.